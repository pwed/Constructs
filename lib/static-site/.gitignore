*.js
tsconfig.json
*.js.map
!jest.config.js
*.d.ts
node_modules
.jsii

# CDK asset staging directory
.cdk.staging
cdk.out
