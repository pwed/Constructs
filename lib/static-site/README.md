# static-site

Easily deploy a static site backed by s3 and CloudFront

## Usage

``` typescript
import { pwed_static_site } from "pwed-cdk";
import * as path from "path";

const website = new pwed_static_site.StaticSite(this, "PwedDotMe", {
    domain: "pwed.me",
    path: path.join(__dirname, "static"),
})
```
