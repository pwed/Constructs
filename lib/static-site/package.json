{
  "name": "static-site",
  "version": "0.1.0",
  "main": "index.js",
  "types": "index.d.ts"
}
