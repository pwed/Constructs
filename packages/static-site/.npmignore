*.ts
!*.d.ts

# CDK asset staging directory
.cdk.staging
cdk.out


# Exclude jsii outdir
lib

# Include .jsii
!.jsii


# Exclude jsii outdir
dist
