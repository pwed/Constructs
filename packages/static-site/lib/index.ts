import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
// import * as sqs from 'aws-cdk-lib/aws-sqs';

export interface StaticSiteProps {
  readonly name: string;
}

export class StaticSite extends Construct {

  constructor(scope: Construct, id: string, props: StaticSiteProps) {
    super(scope, id);

    
    new Bucket(this, "bucket", {
      bucketName: props.name,
    })
  }
}
