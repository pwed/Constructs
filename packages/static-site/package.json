{
  "name": "static-site",
  "version": "0.1.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "author": {
    "name": "Fred Stoddart",
    "email": "freddiestoddart@gmail.com"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pwed/Constructs.git"
  },
  "license": "ISC",
  "scripts": {
    "build": "jsii",
    "build:watch": "jsii -w",
    "package": "jsii-pacmak",
    "test": "tsc && jest",
    "watch": "tsc -w"
  },
  "devDependencies": {
    "@types/glob": "^7.2.0",
    "@types/jest": "^26.0.10",
    "@types/node": "10.17.27",
    "aws-cdk-lib": "2.7.0",
    "constructs": "^10.0.0",
    "jest": "^26.4.2",
    "ts-jest": "^26.2.0",
    "typescript": "~3.9.7"
  },
  "peerDependencies": {
    "aws-cdk-lib": "2.7.0",
    "constructs": "^10.0.0"
  },
  "bundledDependencies": [
    "glob"
  ],
  "stability": "experimental",
  "jsii": {
    "outdir": "dist",
    "versionFormat": "short",
    "targets": {
      "python": {
        "distName": "pwed-constructs.static-site",
        "module": "pwed-constructs.static-site"
      },
      "go": {
        "moduleName": "github.com/pwed/go-constructs/static-site"
      }
    }
  },
  "dependencies": {
    "glob": "^7.2.0"
  }
}
