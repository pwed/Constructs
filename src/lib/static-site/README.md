# static-site

Easily deploy a static site backed by s3 and CloudFront

## Usage

```typescript
import { pwed_static_site } from 'pwed-cdk';
import * as path from 'path';

const website = new pwed_static_site.StaticSite(this, 'PwedDotMe', {
  domain: 'pwed.me',                     // Required
  path: path.join(__dirname, 'static'),  // Required - path to static files
  alternativeDomains: ['blog.pwed.me'],  // Optional - Other names
  hostedZone: zone,                      // Optional - will lookup using the domain if not provided
  enablePrettyPaths: true,               // Optional - will look for index.html in urls ending with a '/' or with no file type
});
```
