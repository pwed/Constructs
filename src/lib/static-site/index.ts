import * as child_process from 'child_process';
import * as crypto from 'crypto';
import * as fs from 'fs';
import { join } from 'path';
import * as process from 'process';
import {
  aws_certificatemanager,
  aws_cloudfront,
  aws_cloudfront_origins,
  aws_route53,
  aws_route53_targets,
  aws_s3,
  aws_s3_deployment,
  Duration,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { sync as globSync } from 'glob';

export interface StaticSiteProps {
  readonly domain: string;
  readonly alternativeDomains?: string[];
  readonly hostedZone?: aws_route53.IHostedZone;
  readonly alternativeHostedZones?: aws_route53.IHostedZone[];
  readonly path: string;
  readonly enablePrettyPaths?: boolean;
}

export class StaticSite extends Construct {
  distribution: aws_cloudfront.IDistribution;
  constructor(scope: Construct, id: string, props: StaticSiteProps) {
    super(scope, id);

    const bucket = new aws_s3.Bucket(this, 'Bucket', {
      encryption: aws_s3.BucketEncryption.S3_MANAGED,
      lifecycleRules: [
        {
          transitions: [
            {
              storageClass: aws_s3.StorageClass.INTELLIGENT_TIERING,
              transitionAfter: Duration.days(7),
            },
          ],
        },
      ],
    });

    const hostedZone = props.hostedZone
      ? props.hostedZone
      : aws_route53.HostedZone.fromLookup(this, `HostedZone${props.domain}`, {
          domainName: props.domain,
        });

    const alternativeHostedZones = props.alternativeHostedZones
      ? props.alternativeHostedZones
      : props.alternativeDomains
      ? props.alternativeDomains.map((domain) => {
          return aws_route53.HostedZone.fromLookup(
            this,
            `HostedZone${domain}`,
            {
              domainName: domain,
            }
          );
        })
      : undefined;
    let domainValidation = {
      [props.domain]: hostedZone,
    };

    props.alternativeDomains?.map((domain, index) => {
      domainValidation[domain] = alternativeHostedZones![index];
    });

    const certificate = new aws_certificatemanager.Certificate(
      this,
      'Certificate',
      {
        domainName: props.domain,
        subjectAlternativeNames: props.alternativeDomains,
        validation:
          aws_certificatemanager.CertificateValidation.fromDnsMultiZone(
            domainValidation
          ),
      }
    );

    const originAccessIdentity = new aws_cloudfront.OriginAccessIdentity(
      this,
      'OriginAccessIdentity'
    );
    bucket.grantRead(originAccessIdentity);

    let prettyPathHandler: aws_cloudfront.Function;
    if (props.enablePrettyPaths)
      prettyPathHandler = new aws_cloudfront.Function(
        this,
        'PrettyPathHandler',
        {
          code: aws_cloudfront.FunctionCode.fromFile({
            filePath: join(__dirname, 'prettyPathHandler.js'),
          }),
        }
      );

    this.distribution = new aws_cloudfront.Distribution(this, 'Distribution', {
      defaultBehavior: {
        origin: new aws_cloudfront_origins.S3Origin(bucket, {
          originAccessIdentity,
        }),
        allowedMethods: aws_cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
        viewerProtocolPolicy:
          aws_cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        functionAssociations: props.enablePrettyPaths
          ? [
              {
                function: prettyPathHandler!,
                eventType: aws_cloudfront.FunctionEventType.VIEWER_REQUEST,
              },
            ]
          : undefined,
      },
      domainNames: props.alternativeDomains
        ? [props.domain, ...props.alternativeDomains]
        : [props.domain],
      certificate,
      errorResponses: [
        {
          httpStatus: 404,
          responseHttpStatus: 404,
          responsePagePath: '/404.html',
        },
      ],
      defaultRootObject: 'index.html',
      httpVersion: aws_cloudfront.HttpVersion.HTTP2_AND_3,
    });

    new aws_route53.ARecord(this, 'AAliasRecord', {
      recordName: props.domain,
      zone: hostedZone,
      target: aws_route53.RecordTarget.fromAlias(
        new aws_route53_targets.CloudFrontTarget(this.distribution)
      ),
    });
    new aws_route53.AaaaRecord(this, 'AaaaAliasRecord', {
      recordName: props.domain,
      zone: hostedZone,
      target: aws_route53.RecordTarget.fromAlias(
        new aws_route53_targets.CloudFrontTarget(this.distribution)
      ),
    });

    props.alternativeDomains?.map((domain, index) => {
      new aws_route53.ARecord(this, `AAliasRecord${domain}`, {
        recordName: domain,
        zone: alternativeHostedZones![index],
        target: aws_route53.RecordTarget.fromAlias(
          new aws_route53_targets.CloudFrontTarget(this.distribution)
        ),
      });
      new aws_route53.AaaaRecord(this, `AaaaAliasRecord${domain}`, {
        recordName: domain,
        zone: alternativeHostedZones![index],
        target: aws_route53.RecordTarget.fromAlias(
          new aws_route53_targets.CloudFrontTarget(this.distribution)
        ),
      });
    });

    const hashFile = '/.hashfile.json';
    let invalidations: string[] = [hashFile];
    invalidations.push(
      ...compareRemoteToLocal(props.domain, hashFile, props.path)
    );

    new aws_s3_deployment.BucketDeployment(this, 'StaticDeployment', {
      sources: [aws_s3_deployment.Source.asset(props.path)],
      destinationBucket: bucket,
      distribution: this.distribution,
      distributionPaths: invalidations,
    });
  }
}

function getHashes(globPattern: string, dir: string): Map<string, string> {
  let fh: Map<string, string> = new Map();
  const gs = globSync(globPattern, { cwd: dir, nodir: true });
  const pwd = process.cwd();
  process.chdir(dir);
  gs.forEach(function (file: string) {
    const fileBuffer = fs.readFileSync(file, {});
    const hashSum = crypto.createHash('sha256');
    hashSum.update(fileBuffer);
    const hex = hashSum.digest('hex');
    fh.set(file, hex);
  });
  process.chdir(pwd);
  return fh;
}

function getInvalidations(
  oldHashes: Map<string, string>,
  newHashes: Map<string, string>
): string[] {
  let invalidations: string[] = [];
  oldHashes.forEach(function (v, k) {
    if (newHashes.get(k) !== v) {
      invalidations.push(`/${k}`);
    }
  });
  return invalidations;
}

function compareRemoteToLocal(
  domain: string,
  hashFile: string,
  localFolder: string
): string[] {
  let oldHashesJSON: string;
  const newHashes = getHashes('**', localFolder);
  fs.writeFileSync(
    join(localFolder, hashFile),
    JSON.stringify(Object.fromEntries(newHashes))
  );
  try {
    oldHashesJSON = child_process
      .execSync(`curl -s https://${domain}/${hashFile}`)
      .toString();
  } catch (e) {
    console.error(e);
    return ['/*'];
  }
  let oldHashes: Map<string, string>;
  try {
    oldHashes = new Map(Object.entries(JSON.parse(oldHashesJSON!)));
  } catch (e) {
    return ['/*'];
  }
  return getInvalidations(oldHashes, newHashes);
}
