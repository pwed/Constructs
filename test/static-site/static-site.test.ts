import * as path from 'path';
import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as StaticSite from '../../src/lib/static-site/index';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'TestStack');

new StaticSite.StaticSite(stack, 'MyTestConstruct', {
  domain: 'example.com',
  path: path.join(__dirname, 'example.com'),
  hostedZone: cdk.aws_route53.HostedZone.fromHostedZoneAttributes(stack, 'HZ', {
    zoneName: 'example.com',
    hostedZoneId: 'Z01234567EXAMPLE8OOO',
  }),
});
const template = Template.fromStack(stack);

test('Has Bucket', () => {
  template.hasResource('AWS::S3::Bucket', {});
});

test('Has Distribution', () => {
  template.hasResource('AWS::CloudFront::Distribution', {});
});

test('Matches Snapshot', () => {
  expect(template).toMatchSnapshot();
});
