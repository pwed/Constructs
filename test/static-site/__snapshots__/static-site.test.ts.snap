// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
Object {
  "Mappings": Object {
    "AWSCloudFrontPartitionHostedZoneIdMap": Object {
      "aws": Object {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": Object {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f98b78092dcdd31f5e6d47489beb5f804d4835ef86a8085d0a2053cb9ae711da.zip",
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "MyTestConstructStaticDeploymentAwsCliLayerD6D4ECC8",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructBucketC00BBB11",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestConstructBucketC00BBB11",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyTestConstructAAliasRecord0B561DBC": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "MyTestConstructDistribution60637A64",
              "DomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::FindInMap": Array [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              Object {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "Z01234567EXAMPLE8OOO",
        "Name": "example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "MyTestConstructAaaaAliasRecordC1F287A4": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "MyTestConstructDistribution60637A64",
              "DomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::FindInMap": Array [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              Object {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "Z01234567EXAMPLE8OOO",
        "Name": "example.com.",
        "Type": "AAAA",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "MyTestConstructBucketC00BBB11": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 7,
                },
              ],
            },
          ],
        },
        "Tags": Array [
          Object {
            "Key": "aws-cdk:cr-owned:048c852e",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyTestConstructBucketPolicyE5FFA623": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestConstructBucketC00BBB11",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructOriginAccessIdentityDC38A3FC",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructBucketC00BBB11",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestConstructBucketC00BBB11",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructOriginAccessIdentityDC38A3FC",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "MyTestConstructBucketC00BBB11",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "MyTestConstructCertificate41948685": Object {
      "Properties": Object {
        "DomainName": "example.com",
        "DomainValidationOptions": Array [
          Object {
            "DomainName": "example.com",
            "HostedZoneId": "Z01234567EXAMPLE8OOO",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "MyTestConstructDistribution60637A64": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "example.com",
          ],
          "CustomErrorResponses": Array [
            Object {
              "ErrorCode": 404,
              "ResponseCode": 404,
              "ResponsePagePath": "/404.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "TestStackMyTestConstructDistributionOrigin16656F6BD",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2and3",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "MyTestConstructBucketC00BBB11",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestStackMyTestConstructDistributionOrigin16656F6BD",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "MyTestConstructOriginAccessIdentityDC38A3FC",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Ref": "MyTestConstructCertificate41948685",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyTestConstructOriginAccessIdentityDC38A3FC": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "MyTestConstructStaticDeploymentAwsCliLayerD6D4ECC8": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "39ee629321f531fffd853b944b2d6f3fa7b5276431c9a4fd4dc681303ab15080.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "MyTestConstructStaticDeploymentCustomResource1C2467CA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucketName": Object {
          "Ref": "MyTestConstructBucketC00BBB11",
        },
        "DistributionId": Object {
          "Ref": "MyTestConstructDistribution60637A64",
        },
        "DistributionPaths": Array [
          "/.hashfile",
          "/*",
        ],
        "Prune": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": Array [
          "e68716fff4c319c8e3ae7fd5ebbf664504e7548ccee6c0b7da080ece1e5b1fb4.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
